hazelcast/src/main/java/com/hazelcast/internal/iteration/IndexIterationPointer.java

```diff
@@ -0,0 +1,168 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.iteration;
+
+import com.hazelcast.map.impl.MapDataSerializerHook;
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+import com.hazelcast.query.impl.AbstractIndex;
+import com.hazelcast.sql.impl.exec.scan.index.IndexEqualsFilter;
+import com.hazelcast.sql.impl.exec.scan.index.IndexFilter;
+import com.hazelcast.sql.impl.exec.scan.index.IndexInFilter;
+import com.hazelcast.sql.impl.exec.scan.index.IndexRangeFilter;
+import com.hazelcast.sql.impl.expression.ExpressionEvalContext;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class IndexIterationPointer implements IdentifiedDataSerializable {
+
+    private static final byte FLAG_DESCENDING = 1;
+    private static final byte FLAG_FROM_INCLUSIVE = 1 << 1;
+    private static final byte FLAG_TO_INCLUSIVE = 1 << 2;
+    private static final byte FLAG_POINT_LOOKUP = 1 << 3;
+
+    private byte flags;
+    private Comparable<?> from;
+    private Comparable<?> to;
+
+    public IndexIterationPointer() {
+    }
+
+    private IndexIterationPointer(
+            byte flags,
+            Comparable<?> from,
+            Comparable<?> to
+    ) {
+        this.flags = flags;
+        this.from = from;
+        this.to = to;
+    }
+
+    public static IndexIterationPointer create(
+            Comparable<?> from,
+            boolean fromInclusive,
+            Comparable<?> to,
+            boolean toInclusive,
+            boolean descending
+    ) {
+        return new IndexIterationPointer(
+                (byte) ((descending ? FLAG_DESCENDING : 0)
+                        | (fromInclusive ? FLAG_FROM_INCLUSIVE : 0)
+                        | (toInclusive ? FLAG_TO_INCLUSIVE : 0)
+                        | (from == to ? FLAG_POINT_LOOKUP : 0)),
+                from,
+                to);
+    }
+
+    public static IndexIterationPointer[] createFromIndexFilter(IndexFilter indexFilter, ExpressionEvalContext evalContext) {
+        ArrayList<IndexIterationPointer> result = new ArrayList<>();
+        createFromIndexFilterInt(indexFilter, evalContext, result);
+        return result.toArray(new IndexIterationPointer[0]);
+    }
+
+    private static void createFromIndexFilterInt(
+            IndexFilter indexFilter,
+            ExpressionEvalContext evalContext,
+            List<IndexIterationPointer> result
+    ) {
+        if (indexFilter instanceof IndexRangeFilter) {
+            IndexRangeFilter rangeFilter = (IndexRangeFilter) indexFilter;
+
+            Comparable<?> from = null;
+            if (rangeFilter.getFrom() != null) {
+                Comparable<?> fromValue = rangeFilter.getFrom().getValue(evalContext);
+                if (fromValue == null) {
```

Comment:
I think this is worth a comment, e.g.:
> this happens e.g. for `a>null`, which gives no results at all because `>` operand applied to null produces UNKNOWN